### ISSUE TITLE
Garbage collection being blocked?

### ISSUE THREAD

**Issue by Lange** [twitch-irc/twitch-irc#80]
Hey there,

This is related to the memory leak I [mentioned on Twitter](https://twitter.com/VanCamp/status/588526161052897283) a few days ago. The leak itself seems to be have been fixed, but now it seems that garbage collection is never running.

My test setup for analyzing these leaks has been very primitive, so it could be that there is a flaw in my process. In any case, I'll go over the fully history to hopefully shed more light on it:

I've been having memory leak problems on my client's NodeCG servers for some time now. Nobody else running NodeCG seemed to have leaks though, and their RAM usage was quite stable. After turning a bunch of things off one-by-one, I realized that the leak was coming from my [lfg-sublistener](https://github.com/SupportClass/lfg-sublistener) bundle. Further experimentation revealed that it was specifically twitch-irc that was causing the leak. Disabling twitch-irc and relying purely on lfg-sublistener's API polling resulted in stable RAM usage that did not deteriorate over time.

I set up periodic heap dumps using [node-heapdump](https://github.com/bnoordhuis/node-heapdump), and these dumps backed up my hypothesis. Each dump had more and more data from twitch-irc in it, and it seemed that this data never got garbage collected. All the data seemed to point to the `channelUserData` property in the now-removed `data.js` library.

Schmoopie refactored the offending part of the code and it looked like the leak was fixed. I pulled down his changes onto my test server, and the heap dumps looked clean. No rising RAM usage, no huge pile of `channelUserData`.

However, when I turned off the periodic heap dumps, the exact same problem returned. RAM usage started steadily rising while chat was active, and it never went down. Baffled, I turned the heap dumps back on. The problem once again went away. It seems that the act of turning on these heap dumps causes the problem to go away.

I hypothesized that this was due to the heap dumps forcing a garbage collection cycle, so I removed the periodic heap dumps in favor of periodic forced garbage collection via `--expose-gc` and `global.gc()`. Sure enough, this made the problem go away as well. It would seem that the leak itself has been fixed, but Node is for some reason deciding to never run garbage collection.

Below I've attached a screenshot of my server's RAM usage. On the left, you can see the RAM usage of a server running twitch-irc normally, with no heap dumps and no forced garbage collection. The ram usage has been steadily climbing whenever the channel it is listening to goes live and its chat activity picks up. The blue lines are where I'm editing and restarting NodeCG to have periodic forced garbage collection every 30 minutes. After this, RAM usage starts out by rising very fast and peaking at 300MB, but then levels out and stabilizes at about 138MB.

![NodeCG RAM graph](https://cloud.githubusercontent.com/assets/873012/7244037/e5cad016-e79a-11e4-8dc0-ece60e54c26b.png)

I've tried to do research into Node's garbage collection to understand why this might be happening, but so far I've not been able to find anything that would explain this pattern I'm seeing. Was this something you also saw when profiling twitch-irc on Strongloop?

**Comment by Lange** [comment]
I also think that the image schmoopie [showed me on twitter](https://i.imgur.com/N0SH0lC.png) after he fixed the leak is inaccurate. In it, you see that heap and GC are flat and close to zero, yet RSS is continuing to climb. I observed something similar when using StrongOps, and it seems to be caused by turning on heap profiling.

On the left of the graph, we see NodeCG running lfg-sublistener. On the right, I have turned on heap profiling in StrongOps and the graph goes haywire.

![graph](https://cloud.githubusercontent.com/assets/873012/7263781/53ba30bc-e84a-11e4-8ef8-8f4e5be4f0eb.png)

**Comment by Schmoopiie** [comment]
Hello @Lange,

Yes I did notice that and I believe it's related to the event listeners, but I still can't confirm anything yet. I can see the exact same results with 2 servers out of 3 so far, with Strongloop I was able to see the memory leak which is now fixed, but I also saw that the RSS was going up and down, which is normal and depends on the page file, but sometimes it wasn't going down enough. I have a feeling that this is due to an event listener that is including an object and the fact that the object is getting used over again and again, it doesn't get GC'ed properly.

I will need some time to figure this out, for now I'm busy with my job and personal life :/ It is one of my priority though.

Thanks!

**Comment by Lange** [comment]
Yeah it's weird. I've been trying to attack this from all sides to find the root cause, but haven't been successful yet.

My production setups are just barely stable enough to work right now, so this isn't a disaster situation for me. They're running out of RAM every few hours, but they're stable enough to restart themselves with only about a minute of downtime.

I'll keep working away at this as well, and I'll let you know if I discover anything useful.

**Comment by Lange** [comment]
I made a [fork of twitch-irc](https://github.com/Lange/twitch-irc) that is gutted down to just the parts that I need for [lfg-sublistener](https://github.com/SupportClass/lfg-sublistener). The good news is that twitch-irc, when run in a minimal test case, has very stable RAM usage. The bad news is that when run in NodeCG as a whole, something is still slowly leaking.

So I've probably got a leak in my code somewhere, but I'm struggling to find it because of the aforementioned heapdump problem. I'll keep looking.

**Comment by celluj34** [comment]
Is there a temporary workaround for the time being? Maybe force GC every 15-30 minutes?

**Comment by Schmoopiie** [comment]
Been working on this for a while now, this is what I have found so far and everyone is welcome to replicate my steps to confirm the results. I tested everything on Windows and Ubuntu 14.04.

Many times I tried using profilers and many memory management applications for Node and as soon as I started profiling, the RSS went up like crazy and this is a problem for me when my goal is to see if there is a memory leak or not. The only memory leak I was able to fix was the ``channelUserData`` thing, but now we are talking about the RSS going up, which might be due to another memory leak. I decided to test with different modules, I tried using a very basic express application and another one with node-irc, both of them started with 40mb RSS and when I started the profiler, they increased to 170mb RSS. Anyways, I used node-heapdump and started creating heapdumps every 60 secs and found out that it might be because of [this function](https://github.com/twitch-irc/twitch-irc/blob/4481991ea05f97b79676ba9ad56b258bd917a944/libraries/client.js#L96). Everywhere in the memory, when there was an array somewhere, this function was added to it. I removed it and changed the ``mods.clean('');`` line in client.js to a for loop and ran the script again and checked the heapdump to make sure this function is no longer being used. I was also looking for data not being garbage collected, but found nothing in about 20-25 different heapdump files. Now that it was gone, I removed node-heapdump and started profiling my own way. I'm using this in my code in client.js and it is still running as I am writing this post..

```javascript
function format (val){
    return Math.round(((val / 1024 / 1024) * 1000) / 1000) + 'mb';
}

setInterval(function () {
    console.log('Memory Usage :: ' +
            ("\tRSS: " + format(process.memoryUsage().rss)) +
            ("\tHeap Total: " + format(process.memoryUsage().heapTotal)) +
            ("\tHeap Used: " + format(process.memoryUsage().heapUsed))
    );
}, 5000);
```
I also tried running the script with ``--trace-gc`` and I can confirm that twitch-irc gets GC'ed.. [the delay is just higher with Node v0.12](https://strongloop.com/strongblog/performance-node-js-v-0-12-whats-new/). Node is garbage collecting twitch-irc every 10mins or so, the first GC increases the RSS a bit and then it is stable, the heap used increases until it gets GC'ed and goes back down. The following screenshot is an instance that has been running the module for 72 hours on 15 channels, it is not running with strongloop or any profilers, just the code I provided above..

![](http://i.imgur.com/rVIrKjE.jpg)

"Heap used" goes up and down as you can see and RSS isn't that high for an application that is analyzing so many lines right now (LOT of chat activity). Everytime there is a message on Twitch, the module receives 4-5 messages. Twitch sends informations with jtv before every message and the module throw it back to you in the user object of the chat and action listeners. Also, Node is single threaded, which might be the cause of high memory consumption when being on many many channels.

The "Heap used" was going up and up while writing this post, RSS is going up too and then suddenly, see what happened:

![](http://i.imgur.com/CqrtPum.jpg)

I will still run my tests on different servers, but maybe it is fixed now. I am not closing this issue as long as it is not confirmed for good.
